### PR TITLE
Add ubuntu 20.04 (Work for 18.04 too)

### PR DESCRIPTION
**Pull request by Hinara**
Ubuntu 20.04 will be release soon,
xrdp had some update making that you SHOULD NOT use `use_vsock=true` anymore
See https://github.com/neutrinolabs/xrdp/issues/1403

### CODE REVIEW

**Comment by sn0wf4ke** [comment]
Thanks. It is working fine. Used it here for my vagrant box. https://app.vagrantup.com/sn0wf4k3/boxes/ubuntu-focal-desktop

**Comment by Hinara** [comment]
This PR is also able to fix new version of the xrdp server that has been released recently on 18.04

**Comment by mariuszkochanowski** [comment]
I used optimized ubuntu 20.04 vhd image from
https://github.com/CanonicalLtd/ubuntu-desktop-hyper-v/tree/master/HyperVGallery

and enhanced session didn't work. 

I uninstalled xorgxrdp, xrdp and removed configuration, then used your script
and it didn't work either.

I made sure that configuration is correct. Xrdp verssion installed was 0.9.12.
Of course i enabled enhanced session on hyperv. 
Older scripts worked for me with ubuntu 18.04 and 19.10 optimized images.

**Comment by maxplatonov** [comment]
Works fine after upgrading an existing 18.04 VM to a 20.04. 
After running the `install.sh` I had to run this in Powershell:
```
Set-VM -VMName <your_vm_name> -EnhancedSessionTransportType HvSocket
```
👍

**Comment by Hinara** [comment]
@mariuszkochanowski Check that xrdp is running correctly `systemctl status xrdp` and that you correctly set that your vm support Enhanced Session with the command just above

**Comment by mariuszkochanowski** [comment]
It works only with ubuntu installed from iso. It doesn't work on ubuntu optimized image - it has some issues.

**Comment by Hinara** [comment]
Xrdp seems to be configured correctly, and manage to launch without crashing (`systemctl status xrdp` report that the process is running)
I think there is definitively something wrong with the optimized image and microsoft modules as even shared-drives is not created in the home folder. So as long as this is not fixed, I can't really do anything, sorry,

**Comment by hiyangw** [comment]
It's working for me if create a ubuntu 20.04 VM manually. Thanks

**Comment by wittelw** [comment]
Like others above, confirmed it does NOT work with the April 23, 2020 Quick Create Hyper-V image. Installed manually and running the following (from above and other sources).

> Set-VM -VMName <your_vm_name> -EnhancedSessionTransportType HvSocket

@Hinara, thanks for this, a much needed PR. Perhaps @cwilhit can comment on status.

Note: I believe the wiki at https://github.com/Microsoft/linux-vm-tools/wiki/Onboarding:-Ubuntu should be updated at the same time as well as changing the version info comment on line 4 of install.sh.

File: ubuntu/20.04/install.sh [hunk 1]
```diff
@@ -0,0 +1,108 @@
+#!/bin/bash
+
+#
+# This script is for Ubuntu 18.04 Bionic Beaver to download and install XRDP+XORGXRDP via
```
Comment (wittelw) [inline ubuntu/20.04/install.sh:4]: Update the comment to 20.04 version.

**Comment by elvries** [comment]
I needed the updates in http://c-nergy.be/blog/?p=14888 (xrdp-installer-1.2.sh) to solve the rights issues which popup after startup. This was a clean install from the Ubuntu 20.04 iso, so I would say everybody should have these issues. In that case I guess the relevant parts of that script could be copied to this script.

**Comment by Metalscreame** [comment]
Hi all
@elvries  you mean `Can't open PID file /run/xrdp/xrdp.pid (yet?)` error? 
Im getting it after script installation of xrdp. in the `systemctl status xrdp.service`
And enhanced mode is not being enabled. Im not sure if this is the 

I also tried the one you mentioned - but still had the some error

**Comment by elvries** [comment]
@Metalscreame, no, I did not have that issue.

**Comment by Hinara** [comment]
If anyone would like to try again with a quick create version ?
Could not reproduce this issue by myself anymore

**Comment by srcnet2** [comment]
It does not work still

**Comment by Hinara** [comment]
Try to shutdown VM and them restart it ?
Maybe windows version ? 1909 OS build: 18363.815
With quick create, it works :/

**Comment by srcnet2** [comment]
It is working now, but I don't know how or what I did to fix it. I rebooted the host, the guest. I ended up saving the guest at one point because I was going to delete it. 

Then it started working

**Comment by mykter** [comment]
Didn't work for me with a quick-create image from today, on build 17763. Script runs fine, set the VM session transport type, but it still gets the normal unenhanced session. There were no obvious xrdp errors in the logs, but I'm not sure what else I should be looking for.

**Comment by grobius** [comment]
I only get is to work when first starting it up and then choose save state and described by srcnet2

**Comment by ImportTaste** [comment]
The Save method worked for me as well.

**Comment by mykter** [comment]
Saving and then resuming at the logon prompt did it, thanks!

**Comment by matthiasbalke** [comment]
Your install script works fine for me, using the quick create on Windows 1909! Thx!

**Comment by eckorog2005** [comment]
I had success today with quick create on Windows 1909 after saving the VM and resuming again.

**Comment by RvdHout** [comment]
@Hinara Hinara
I got enhanced session mode working using install.sh, but unfortunately i now can no longer use sudo with, for example:

**sudo gedit**
```
No protocol specified
Unable to init server: Could not connect: Connection refused

(gedit:2075): Gtk-WARNING **: 11:30:09.960: cannot open display: :10.0
```

**sudo nautilus**
```
** (org.gnome.Nautilus:1884): WARNING **: 11:19:19.186: Error on getting connection: Failed to load SPARQL backend: Error spawning command line ?dbus-launch --autolaunch=3e96575b2ac04618bf7fbd9083666c53 --binary-syntax --close-stderr?: Child process exited with code 1
No protocol specified
Unable to init server: Could not connect: Connection refused

(org.gnome.Nautilus:1884): Gtk-WARNING **: 11:19:19.194: cannot open display: :10.0
```

This seems no problem in  ubuntu 18.x, can it be related to no longer using use_vsock? Anyone else can confirm these errors or knows how to fix them?

**Comment by Hinara** [comment]
@RvdHout I assume that this is not from xrdp but that xorg configuration might might have been made me more strict.
For use_vsock you can try too put it to true but as you can see in the liked issue it seems that this flag is now obsolete and might not be used at all, so I don't think it is the root of your issue.

**Comment by RvdHout** [comment]
> @RvdHout I assume that this is not from xrdp but that xorg configuration might might have been made me more strict.
> For `use_vsock` you can try too put it to true but as you can see in the liked issue it seems that this flag is now obsolete and might not be used at all, so I don't think it is the root of your issue.

Strange thing is that i can both execute gedit and nautilus using sodo on 20.04 without enhanced session mode enabled successfully, so somewhere it looks related

**Comment by Hinara** [comment]
As said before this has nothing to do with xorg x)
So to explain it xrdp give a remote access to an xorg session which depending on the configuration might forbid some things for remote session.
And I think I may be wrong to because `sudo` works, but you need to use `sudo -E` as some environment variables might not be passed correctly to the root user by sudo preventing the connection to xorg as root.
TLDR: Use `sudo -E`
If anybody find a way that allow to do the same thing without using `-E` with sudo, please do a PR on my fork ;)

**Comment by RvdHout** [comment]
sudo -E seems to work (partially.... i get a bunch of (error)messages in terminal but gedit executes OK)

**Comment by wittelw** [comment]
Can confirm @Hinara suggestion of sudo -E on 20.04 (mine was from ISO rather than Hyper-V Quick Connect image. I also have a 18.04.3 VM installed using Quick Connect image, and that VM does not require the -E sudo param. And of course 20.04 on bare metal doesn't need it either.

Clearly some things are not working as seamlessly on 20.04 in Hyper-V as on 18.04.3, but lean towards agreeing these problems are likely in Ubuntu 20.04 vs. this pull request, which at least allows 20.04 to run in Hyper-V (with the various tweaks mentioned above).

If anyone that worked on the 18.04.x Quick Connect image is listening, perhaps they can provide some insight.

**Comment by ImportTaste** [comment]
add `Defaults env_keep += "DISPLAY HOME"` to /etc/sudoers to fix the 'No protocol specified, cannot open display' issue.

Use the visudo tool if you have to.

**Comment by RvdHout** [comment]
+1

**Comment by RvdHout** [comment]
> add `Defaults env_keep += "DISPLAY HOME"` to /etc/sudoers to fix the 'No protocol specified, cannot open display' issue.
> 
> Use the visudo tool if you have to.
 👍

**Comment by Hinara** [comment]
`DISPLAY` is already kept by default but `HOME` is not, strange that xorg rely on this but well ^^"

**Comment by ImportTaste** [comment]
@Hinara I recommend adding a line to append it to the sudoers file to this PR for those happening across this for the first time.

**Comment by luisaranda** [comment]
Weird....Still it isn't working for me. I've tried the Save method and the PowerShell command

**Comment by ImportTaste** [comment]
> Weird....Still it isn't working for me. I've tried the Save method and the PowerShell command

Make sure you actually saved /etc/sudoers because it will try to make the filename /etc/sudoers.tmp if you use the visudo tool

**Comment by AntonOfTheWoods** [comment]
Sorry, I'm just getting started again with Hyper-v after something like 10 years... I pulled this on the original branch (branch20-04) and most things look like they are working. I get xrdp, and copy/paste is working fine. The `shared-drives` directory in the file explorer on the guest is empty though - is there something extra I need to do on the host?

**Comment by Hinara** [comment]
I was checking something, I'm not sure if you should modify /etc/sudoers to keep the HOME variable.
It might interfere with some script that try to use current user home folder as bash use the HOME environment variable to locate current user HOME folder.

**Comment by ImportTaste** [comment]
I guess the thing to do then is put all the user's HOME files into /root and then systematically remove them until it's figured out which one is required.

**Comment by irfannaseefp** [comment]
Any updates on this ? I couldnt make it work using this install.sh script. systemctl shows xrdp is running, but xrdp is not listening on any port.

**Comment by Hinara** [comment]
@irfannaseefp did you try the solution above https://github.com/microsoft/linux-vm-tools/pull/106#issuecomment-627584514 ?
I can't do anything about this as it seems to be a bug from HyperV. On my computer everything works great without touching anything, I can't reproduce this anymore, a windows update might have fix this.

**Comment by oldium** [comment]
Tested, works good. One annoyance, though - I have to enter password after first login to allow package auto-update and also when I want to reboot computer. There is a fix for this in Xubuntu pull request #113 in the [created `xrdp-allow-all.pkla` file](https://github.com/microsoft/linux-vm-tools/pull/113/commits/2282027e99c0e56b0ebb215f562b632bd52a8123#diff-8e1a7b7043b6c02a7bcca4f7550cd80eR91-R104). It would be nice to have it for Ubuntu too I think (just my personal opinion).

**Comment by lorenzobilli** [comment]
@oldium I've originally created the aforementioned xrdp-allow-all.pkla for my Kubuntu pull request. I'm testing it on the vanilla Ubuntu 20.04, and it seems working well. No strange authorization popup issues. So yeah, I think it could be integrated also in the vanilla Ubuntu install script.

**Comment by trailstrider** [comment]
Still not any easy way to get this working with 20.04 LTS it seems.  After trying script and a couple approaches, none have worked.

**Comment by oldium** [comment]
I recommend creating new virtual machine, enable Integration Services/Guest Services in VM properties. Then install Ubuntu 20.04 DVD (I used `ubuntu-20.04-desktop-amd64.iso`). Then 

```bash
sudo apt install git
git clone https://github.com/microsoft/linux-vm-tools.git
cd linux-vm-tools
git fetch --all
git pull origin pull/106/head:pull_106
git checkout pull_106
cd ubuntu/20.04/
sudo bash ./install.sh
```

You will need to reboot and repeat `sudo bash ./install.sh`. Then one more restart and you can try the Enhanced Session.

**Comment by luisaranda** [comment]
So it appears to be an XRDP bug this will be patched in the quick create gallery once 20.04.1 is released. See for more info https://twitter.com/unixterminal/status/1268918077141528579

**Comment by SasukeeUchiha** [comment]
Why isn't this merged yet?

**Comment by cprivite** [comment]
For those who go through all of this stuff (you set the vsock settings, disabled auto login, and rebooted everything) and still don't get the enhanced session, something I found is that you need to tell hyper-v to try to start the Enhanced Session AFTER xrdp is started up. Two ways to do this:
1) As mentioned earlier in the thread, save the VM state while at the login prompt and then resume it. That'll get it to try enhanced mode.
2) There is a button in the Virtual Machine window that toggles between enhanced and basic sessions. You can just click that to switch to an enhanced session.

**Comment by relief-melone** [comment]
Why is this still not merged!! Microsoft could you please wake up! Works like a charm

**Comment by smailc** [comment]
Is this going to be merged?

**Comment by dcui** [comment]
FYI:
I followed
https://askubuntu.com/questions/1246362/how-to-setup-xrdp-for-ubuntu-20-04-in-windows-hyper-v
https://medium.com/@francescotonini/how-to-install-ubuntu-20-04-on-hyper-v-with-enhanced-session-b20a269a5fa7
and I was able to enable xrdp in my Ubuntu 20.04 VM. Basically I did "wget https://raw.githubusercontent.com/Hinara/linux-vm-tools/ubuntu20-04/ubuntu/20.04/install.sh;
sudo chmod +x install.sh;
sudo ./install.sh", and 
"Set-VM -VMName <your_vm_name> -EnhancedSessionTransportType HvSocket" (this runs on the host).

However, after I logged in via xrdp I only got an empty screen... "ps aux |grep -i xorg" showed that 2 Xorg servers were running, so I disabled the "traditional" Xorg server via "systemctl set-default multi-user.targe; reboot", and now I'm able to see the normal desktop after I login via xrdp. :-)

**Comment by gmnowels** [comment]
> FYI:
> I followed
> https://askubuntu.com/questions/1246362/how-to-setup-xrdp-for-ubuntu-20-04-in-windows-hyper-v
> https://medium.com/@francescotonini/how-to-install-ubuntu-20-04-on-hyper-v-with-enhanced-session-b20a269a5fa7
> and I was able to enable xrdp in my Ubuntu 20.04 VM. Basically I did "wget https://raw.githubusercontent.com/Hinara/linux-vm-tools/ubuntu20-04/ubuntu/20.04/install.sh;
> sudo chmod +x install.sh;
> sudo ./install.sh", and
> "Set-VM -VMName <your_vm_name> -EnhancedSessionTransportType HvSocket" (this runs on the host).
> However, after I logged in via xrdp I only got an empty screen... "ps aux |grep -i xorg" showed that 2 Xorg servers were running, so I disabled the "traditional" Xorg server via "systemctl set-default multi-user.target; reboot", and now I'm able to see the normal desktop after I login via xrdp. :-)

I'm running Ubuntu gnome desktop. Got empty screen initially as well. Simply rebooted.

**Comment by Hinara** [comment]
The actual configuration script does not touch Xorg at a single point so if it doesn't work I suggest you to check first if there might be an issue with Xorg & systemd or Xorg & xrdp that makes two servers run at the same time.
This script does just install the virtualized kernel variation that have some kernel module by default inside
Check if there is no update to other packages.
Install xrdp and configure it
Configure over utilities to work with remote connection.
Ask you to reboot.

**Comment by TTimo** [comment]
This isn't working for me ..

The `install.sh` script executes without error, I shutdown the VM, issue the `Set-VM` command to set `HvSocket`, reboot ..

The VM comes back online, it's waiting at the gdm greeter for my user password, there is no prompt for resolution when I connect, no xrdp login screen.

`systemctl status xrdp` `systemctl status xrdp-sesman` report good state, `Xorg` is running ..

**Comment by TTimo** [comment]
Nevermind! It took a reboot of the Windows host, which I hadn't considered .. those things happen pretty often.

**Comment by okand** [comment]
This is probably unsupported so never mind me but I still want to write it down here.

I tried this using pop_os 20.04 and after rebooting the windows host and running `Set-VM -VMName pop_os  -EnhancedSessionTransportType HvSocket` I got hyper-v to connect with an enhanced session and it showed me the xrdp login screen. After logging in I just get a fullscreen error message that's as vague as Microsofts bsods these days and just says "something" went wrong and that I need to log out and try again/something else though.

**Comment by Hinara** [comment]
Blame microsoft drivers for that not something I can do something for :/ as it's a bug on windows side not on VM side

**Comment by okand** [comment]
It's inside the VM the error appears though. Looks like this: https://okand.website/ohno.png

**Comment by Hinara** [comment]
Ohh sry was thinking windows did that not the VM.
Is it only your session our the whole VM ?
If it's the VM  not sure what exactly can cause that Linux Kernel doing something definitively wrong crashing like that but seems strange that Hyper-V stop that as you can't even get crash informations.
Hyper-V state get corrupted and Hyper-V shutdown everything to prevent further damages ?
If it's the RDP session might need to make a ticket to xrdp project as it might be some invalid data.

**Comment by okand** [comment]
I just did an install with regular ubuntu 20.04 instead and that works fine. It's something specific to pop_os then, more things different with it than just being ubuntu but without canonical I suppose.

**Comment by platima** [comment]
Oh this has had me so stumped.
Windows 10 build 1909; quick create, run script, reboot, run script, shutdown, set EnhancedSessionTransportType HvSocket, boot, no bueno. Tried with and without vsock, confirmed xrdp running fine, confirmed only 1 xorg running, confirmed kernel modules correct, confirmed that the 'Enhanced Session' option was definitely not available, tried with and without both Secure Boot and Guest Integration Services.
Tried again with ubuntu-20.04.1-desktop-amd64.iso installer, minimal, same process; still no luck. Then also tried `systemctl set-default multi-user.target` as a moonshot but that didn't help.
I _haven't_ tried rebooting the host yet, as I have a pile of work open right now, but I really cannot think how that would help as this worked with the 18.04 install.

**Comment by AntonOfTheWoods** [comment]
Actually the machine I had installed these on broke somehow (the two NICs I'd added got broke, I tried to boot to recovery but the console wouldn't launch, meaning no way of recovering anything short of mounting and doing a chroot).

So what is the point of these tools again? 

At least for Ubuntu, it is completely obvious why MS is totally MIA on this - WSL2. Sure, they haven't done desktop apps yet but they probably want to focus on doing that rather than wasting time here. In addition, you still don't get sound with these and copy paste works fine with a normal `xrdp` install and `mstsc`. The auto-mounting of local drives is more trouble than it's worth - I regularly had it not mount them so if you have anything that points to them and auto-creates directories you get stuff inside the VM rather than on the host. 

My advice? Forget about these "tools" and:

- install `xrdp`
- share any directories you want on the host and mount them like normal samba drive using `cifs-utils` (same problems with mounting can occur but not more so)
- install the `azure` kernels from the normal repo

Hyper-V is definitely NOT a robust desktop virtualisation solution, it is a (pretty good) server virtualisation solution that probably also works well for virtualisation of Windows desktops in enterprise environments. For the rest it is significantly inferior to the competition, including free variants like VirtualBox. The only reason not to disable it on Win10 is because that means you have to completely disable WSL2. Having it activated breaks all other good desktop virtualisation software, so for the moment looks like we are stuck with it. When WSL2 does GUI apps, it will be (almost) totally obsolete for virtualising Linux on Win10 (but will still be required...).

**Comment by bonzani** [comment]
works well +1

**Comment by koder101** [comment]
I could confirm that this PR-106 works with Ubuntu-20.10 as well.

**Comment by jeremyn** [comment]
@koder101's summary comment at https://github.com/microsoft/linux-vm-tools/pull/106#issuecomment-751276679 worked for me with Ubuntu 20.04 if I installed using an ISO, not "quick create". I also think I needed to reboot the host. After it was all done the guest graphics were still noticeably laggier than a similar setup in VMware Workstation Pro 16.

I agree with @AntonOfTheWoods at https://github.com/microsoft/linux-vm-tools/pull/106#issuecomment-713250529: Hyper-V is not the best option for desktop virtualization, like if you want to run a full graphical Linux desktop side-by-side with Windows. By the way, with VMware Workstation Pro all I needed to do to get Hyper-V and VMware working side-by-side was enable the "Windows Hypervisor Platform" optional feature in Windows 10 Pro, which VMware prompts you to do during its install. It should be the same for VirtualBox but I haven't tried it and I'm seeing mixed reports so YMMV there.

**Comment by iancward** [comment]
I can confirm that this does not work with an Ubuntu 20.04 quickstart VM (not entirely sure why), but does with with a fresh Lubuntu (Ubuntu with LXDE) install from the mini.iso netinstall image. I had to fully stop the VM and restart it in order it to switch to "Enhanced Mode" but it works great now.

Thanks!

**Comment by DRCHRIS2** [comment]
Bump... works great with a vanilla 20.10 install from ISO. Clone the PR, install, reboot, re-run install per _stdout_, shutdown, adjust the VM config, restart VM and it's flawless. A little circuitous but still good. MS should really consider spending some time on these PRs.

**Comment by platima** [comment]
I found that nothing has to be done with quick start VMs now. Not sure if MS updated, but deploy 20 from quickstart, reboot, and it's enhanced.

**Comment by dboreham** [comment]
> I found that nothing has to be done with quick start VMs now. Not sure if MS updated, but deploy 20 from quickstart, reboot, and it's enhanced.

Reboot the guest or the host? (not working for me fwiw)
